lib/membrane_tcp/socket.ex

```diff
@@ -19,34 +17,49 @@ defmodule Membrane.TCP.Socket do
           connection_side: :server | :client | {:client, :inet.ip_address(), :inet.port_number()},
           local_address: :inet.socket_address(),
           local_port_no: :inet.port_number(),
-          local_socket: t() | nil
+          local_socket: :gen_tcp.socket() | nil
         }
 
   @spec create_socket_pair(socket_pair_config(), keyword()) ::
           {local_socket :: t(), remote_socket :: t() | nil}
-  def create_socket_pair(
-        %{connection_side: connection_side, local_socket: local_socket} = sockets_config,
-        local_socket_options \\ []
-      ) do
+  def create_socket_pair(sockets_config, local_socket_options \\ []) do
     local_socket =
-      case local_socket do
+      case sockets_config.local_socket do
         nil ->
           %__MODULE__{
             ip_address: sockets_config.local_address,
             port_no: sockets_config.local_port_no,
             sock_opts: local_socket_options,
-            connection_side: connection_side
+            connection_side: sockets_config.connection_side
           }
 
-        %__MODULE__{connection_side: ^connection_side, state: :connected} ->
-          local_socket
+        socket_handle ->
+          {:ok, {socket_address, socket_port}} = :inet.sockname(socket_handle)
+
+          if sockets_config.local_address not in [socket_address, :any] do
+            raise "Local address passed in options not mathing the one of the passed socket."
+          end
+
+          if sockets_config.local_port_no not in [socket_port, 0] do
+            raise "Local port passed in options not mathing the one of the passed socket."
+          end
 
-        _not_matching_connection_side_socket ->
-          raise "Connection side of provided socket not matching options"
+          if not match?({:ok, _peername}, :inet.peername(socket_handle)) do
+            raise "Local socket not connected."
+          end
```

Comment:
[NIT] You could aggregate these statements into a single `cond do ... end` clause